text/3742-mutate-non-mutable-locals.md

```diff
@@ -0,0 +1,252 @@
+```rs
+fn get_num() -> i32 {
+    let x = 12;
+    x = 13; // WARNING! Assigned twice to immutable variable `x`
+    x
+}
+```
+
+Creating a `&mut` reference is also considered a mutation, since a mutation could occur through the reference:
+```rs
+fn get_num() -> i32 {
+    let x = 12;
+    let px = &mut x; // WARNING! Borrowing `x` as mutable when it is not declared as mutable
+    *px = 13;
+    x
+}
+```
+
+`mut` annotations may be applied to other forms of bindings as well, and the same rules apply:
+```rs
+fn do_something(mut arg: Option<i32>){ // Function parameter with `mut`
+    let (mut a, b) = (12, 13);         // Irrefutable pattern with `mut`
+    if let Some(mut val) = arg {}      // Refutable pattern with `mut`
+}
+```
+
+# Reference-level explanation
+[reference-level-explanation]: #reference-level-explanation
+
+The borrow checker has to deal with the `mut` vs `&mut` distinction, and therefore
+already has special handling for locals.  Making immutable re-assignment a lint just
+means converting local-specific code paths in borrowck to trigger lints instead of fatal
+diagnostics.  Because this is a relaxation of existing language semantics,
+the change is fully backwards-compatible, and making the lint deny-by-default means even code which _fails_
+to compile will continue to do so after the change. 
+
+One interesting consequence of this change is that it allows mutation in places where there
+previously was no syntax to allow it, for example in `ref` bindings:
+
+```rs
+let ref x = 12;
+x = x; // Hard error currently, would become lint.
+
+let ref mut x = 12;
+x = x; // Also a hard error currently, would also become lint.
+```
+
+Note: [Upcoming changes to match ergonomics](https://github.com/rust-lang/rust/issues/123076) also enable this.
+
+# Drawbacks
+[drawbacks]: #drawbacks
+
+Most languages with the concept of immutability treat mutation of immutable locals as a hard error, so the
+current behavior is consistent with that.  There are reasons to treat Rust as a special case, though:
+ - Some of the other languages with immutability treat immutable locals as compile-time evaluable,
+   but rust handles that with a distinct concept (`const`)
+ - Rust is "immutable by default," and intentionally makes mutable locals the "noisier" option.  This makes
+   the hard error more ergonomically costly in Rust than it would be in other languages, since it comes
+   up more frequently in practice
+
+Converting the hard error to a lint, even a deny-by-default one, gives people the option to turn
+the lint off.  While Rust doesn't have a culture of blanket disabling lints, people coming from
+other languages with default-mutability may find it a tempting option.  On the other hand, a
+compilation-blocking error hard-coded into rustc is about the most forceful way possible to
+request that developers use a particular coding style, and it's not clear whether that's
+warranted in this case.
+
+# Rationale and alternatives
```

Comment:
This section should cover why this proposal is better than simply removing the use of `mut` for local variables over an edition boundary.